File: app/src/main/java/com/elementary/tasks/creators/fragments/LocationOutFragment.java
package com.elementary.tasks.creators.fragments;

import android.app.Activity;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;

import com.elementary.tasks.R;
import com.elementary.tasks.core.controller.EventControl;
import com.elementary.tasks.core.controller.EventControlImpl;
import com.elementary.tasks.core.fragments.MapFragment;
import com.elementary.tasks.core.interfaces.MapCallback;
import com.elementary.tasks.core.interfaces.MapListener;
import com.elementary.tasks.core.location.LocationTracker;
import com.elementary.tasks.core.utils.Constants;
import com.elementary.tasks.core.utils.LogUtil;
import com.elementary.tasks.core.utils.Permissions;
import com.elementary.tasks.core.utils.Prefs;
import com.elementary.tasks.core.utils.RealmDb;
import com.elementary.tasks.core.utils.SuperUtil;
import com.elementary.tasks.core.utils.ThemeUtil;
import com.elementary.tasks.core.utils.TimeUtil;
import com.elementary.tasks.core.utils.ViewUtils;
import com.elementary.tasks.core.views.ActionView;
import com.elementary.tasks.databinding.FragmentReminderLocationOutBinding;
import com.elementary.tasks.reminder.models.Place;
import com.elementary.tasks.reminder.models.Reminder;
import com.google.android.gms.maps.model.LatLng;

import java.util.Collections;
import java.util.List;

/**
 * Copyright 2016 Nazar Suhovich
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

public class LocationOutFragment extends RadiusTypeFragment {

    private static final String TAG = "DateFragment";
    private static final int CONTACTS = 132;
    private static final int LOCATION = 202;
    public static final int CONTACTS_ACTION = 133;

    private FragmentReminderLocationOutBinding binding;
    private MapFragment mapFragment;

    private LocationTracker mTracker;
    private LatLng lastPos;

    private ActionView.OnActionListener mActionListener = new ActionView.OnActionListener() {
        @Override
        public void onActionChange(boolean hasAction) {
            if (!hasAction) {
                mInterface.setEventHint(getString(R.string.remind_me));
                mInterface.setHasAutoExtra(false, null);
            }
        }

        @Override
        public void onTypeChange(boolean isMessageType) {
            if (isMessageType) {
                mInterface.setEventHint(getString(R.string.message));
                mInterface.setHasAutoExtra(true, getString(R.string.enable_sending_sms_automatically));
            } else {
                mInterface.setEventHint(getString(R.string.remind_me));
                mInterface.setHasAutoExtra(true, getString(R.string.enable_making_phone_calls_automatically));
            }
        }
    };
    private MapCallback mCallback = new MapCallback() {
        @Override
        public void onMapReady() {
            if (mInterface.getReminder() != null) {
                Reminder item = mInterface.getReminder();
                String text = item.getSummary();
                Place jPlace = item.getPlaces().get(0);
                double latitude = jPlace.getLatitude();
                double longitude = jPlace.getLongitude();
                radius = jPlace.getRadius();
                if (mapFragment != null) {
                    mapFragment.setMarkerRadius(radius);
                    lastPos = new LatLng(latitude, longitude);
                    mapFragment.addMarker(lastPos, text, true, true, radius);
//                    toggleMap();
                }
                binding.mapLocation.setText(SuperUtil.getAddress(lastPos.latitude, lastPos.longitude));
                binding.mapCheck.setChecked(true);
            }
        }
    };
    private MapListener mListener = new MapListener() {
        @Override
        public void placeChanged(LatLng place, String address) {
            lastPos = place;
            String location = SuperUtil.getAddress(place.latitude, place.longitude);
            binding.currentLocation.setText(location);
        }

        @Override
        public void onZoomClick(boolean isFull) {
            mInterface.setFullScreenMode(isFull);
        }

        @Override
        public void onBackClick() {
            if (mapFragment.isFullscreen()) {
                mapFragment.setFullscreen(false);
                mInterface.setFullScreenMode(true);
            }
            ViewUtils.fadeOutAnimation(binding.mapContainer);
            ViewUtils.fadeInAnimation(binding.specsContainer);
        }
    };
    private LocationTracker.Callback mLocationCallback = new LocationTracker.Callback() {
        @Override
        public void onChange(double lat, double lon) {
            lastPos = new LatLng(lat, lon);
            String _Location = SuperUtil.getAddress(lat, lon);
            String text = mInterface.getSummary();
            if (TextUtils.isEmpty(text)) text = _Location;
            binding.currentLocation.setText(_Location);
            if (mapFragment != null) {
                mapFragment.addMarker(lastPos, text, true, true, radius);
            }
        }
    };

    @Override
    protected void recreateMarker() {
        mapFragment.recreateMarker(radius);
    }

    @Override
    public boolean save() {
        super.save();
        if (mInterface == null) return false;
        Reminder reminder = mInterface.getReminder();
        int type = Reminder.BY_OUT;
        boolean isAction = binding.actionView.hasAction();
        if (TextUtils.isEmpty(mInterface.getSummary()) && !isAction) {
            mInterface.showSnackbar(getString(R.string.task_summary_is_empty));
            return false;
        }
        if (lastPos == null) {
            mInterface.showSnackbar(getString(R.string.you_dont_select_place));
            return false;
        }
        String number = null;
        if (isAction) {
            number = binding.actionView.getNumber();
            if (TextUtils.isEmpty(number)) {
                mInterface.showSnackbar(getString(R.string.you_dont_insert_number));
                return false;
            }
            if (binding.actionView.getType() == ActionView.TYPE_CALL) {
                type = Reminder.BY_OUT_CALL;
            } else {
                type = Reminder.BY_OUT_SMS;
            }
        }
        if (reminder == null) {
            reminder = new Reminder();
        }
        Place place = new Place(radius, mapFragment.getMarkerStyle(), lastPos.latitude, lastPos.longitude, mInterface.getSummary(), number, null);
        List<Place> places = Collections.singletonList(place);
        reminder.setPlaces(places);
        reminder.setTarget(number);
        reminder.setType(type);
        reminder.setExportToCalendar(false);
        reminder.setExportToTasks(false);
        reminder.setClear(mInterface);
        if (binding.attackDelay.isChecked()) {
            long startTime = binding.dateView.getDateTime();
            reminder.setStartTime(TimeUtil.getGmtFromDateTime(startTime));
            reminder.setEventTime(TimeUtil.getGmtFromDateTime(startTime));
            LogUtil.d(TAG, "EVENT_TIME " + TimeUtil.getFullDateTime(startTime, true, true));
        } else {
            reminder.setEventTime(null);
            reminder.setStartTime(null);
        }
        RealmDb.getInstance().saveObject(reminder);
        EventControl control = EventControlImpl.getController(mContext, reminder);
        control.start();
        return true;
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        setHasOptionsMenu(true);
    }

    @Override
    public void onCreateOptionsMenu(final Menu menu, final MenuInflater inflater) {
        inflater.inflate(R.menu.fragment_location_menu, menu);
        super.onCreateOptionsMenu(menu, inflater);
    }

    @Override
    public boolean onOptionsItemSelected(final MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_custom_radius:
                showRadiusPickerDialog();
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        binding = FragmentReminderLocationOutBinding.inflate(inflater, container, false);
        mapFragment = MapFragment.newInstance(true, true, true, true,
                Prefs.getInstance(mContext).getMarkerStyle(), ThemeUtil.getInstance(mContext).isDark());
        mapFragment.setListener(mListener);
        mapFragment.setCallback(mCallback);
        getFragmentManager().beginTransaction()
                .replace(binding.mapFrame.getId(), mapFragment)
                .addToBackStack(null)
                .commit();
        binding.actionView.setListener(mActionListener);
        binding.actionView.setActivity(getActivity());
        binding.actionView.setContactClickListener(view -> selectContact());

        binding.delayLayout.setVisibility(View.GONE);
        binding.mapContainer.setVisibility(View.GONE);
        binding.attackDelay.setOnCheckedChangeListener((buttonView, isChecked) -> {
            if (isChecked) binding.delayLayout.setVisibility(View.VISIBLE);
            else binding.delayLayout.setVisibility(View.GONE);
        });
        ImageButton mapButton = binding.mapButton;
        if (ThemeUtil.getInstance(mContext).isDark()){
            mapButton.setImageResource(R.drawable.ic_map_white);
        } else {
            mapButton.setImageResource(R.drawable.ic_map);
        }
        mapButton.setOnClickListener(view -> {
            if (binding.mapCheck.isChecked()) {
                toggleMap();
            }
            binding.mapCheck.setChecked(true);
        });
        binding.currentCheck.setOnCheckedChangeListener((compoundButton, b) -> {
            if (binding.currentCheck.isChecked()) {
                binding.mapCheck.setChecked(false);
                if (Permissions.checkPermission(getActivity(), Permissions.ACCESS_COARSE_LOCATION, Permissions.ACCESS_FINE_LOCATION)) {
                    mTracker = new LocationTracker(mContext, mLocationCallback);
                } else {
                    Permissions.requestPermission(getActivity(), LOCATION, Permissions.ACCESS_FINE_LOCATION, Permissions.ACCESS_COARSE_LOCATION);
                }
            }
        });
        binding.mapCheck.setOnCheckedChangeListener((compoundButton, b) -> {
            if (binding.mapCheck.isChecked()) {
                binding.currentCheck.setChecked(false);
                toggleMap();
                if (mTracker != null) {
                    mTracker.removeUpdates();
                }
            }
        });
        binding.currentCheck.setChecked(true);
        editReminder();
        return binding.getRoot();
    }

    private void toggleMap() {
        if (binding.mapContainer != null && binding.mapContainer.getVisibility() == View.VISIBLE) {
            ViewUtils.fadeOutAnimation(binding.mapContainer);
            ViewUtils.fadeInAnimation(binding.specsContainer);
        } else {
            ViewUtils.fadeOutAnimation(binding.specsContainer);
            ViewUtils.fadeInAnimation(binding.mapContainer);
            if (mapFragment != null) {
                mapFragment.showShowcase();
            }
        }
    }

    @Override
    public boolean onBackPressed() {
        return mapFragment == null || mapFragment.onBackPressed();
    }

    private void editReminder() {
        if (mInterface.getReminder() == null) return;
        Reminder reminder = mInterface.getReminder();
        if (reminder.getEventTime() != null) {
            binding.dateView.setDateTime(reminder.getEventTime());
            binding.attackDelay.setChecked(true);
        }
        if (reminder.getTarget() != null) {
            binding.actionView.setAction(true);
            binding.actionView.setNumber(reminder.getTarget());
            if (Reminder.isKind(reminder.getType(), Reminder.Kind.CALL)) {
                binding.actionView.setType(ActionView.TYPE_CALL);
            } else if (Reminder.isKind(reminder.getType(), Reminder.Kind.SMS)) {
                binding.actionView.setType(ActionView.TYPE_MESSAGE);
            }
        }
    }

    private void selectContact() {
        if (Permissions.checkPermission(getActivity(), Permissions.READ_CONTACTS, Permissions.READ_CALLS)) {
            SuperUtil.selectContact(getActivity(), Constants.REQUEST_CODE_CONTACTS);
        } else {
            Permissions.requestPermission(getActivity(), CONTACTS, Permissions.READ_CONTACTS, Permissions.READ_CALLS);
        }
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == Constants.REQUEST_CODE_CONTACTS && resultCode == Activity.RESULT_OK) {
            String number = data.getStringExtra(Constants.SELECTED_CONTACT_NUMBER);
            binding.actionView.setNumber(number);
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        if (mTracker != null) mTracker.removeUpdates();
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        switch (requestCode) {
            case CONTACTS:
                if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    selectContact();
                }
                break;
            case LOCATION:
                if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    binding.currentCheck.setChecked(true);
                }
                break;
            case CONTACTS_ACTION:
                if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    binding.actionView.setAction(true);
                }
                break;
        }
    }
}

File: app/src/main/java/com/elementary/tasks/reminder/createEdit/fragments/LocationOutFragment.kt
package com.elementary.tasks.reminder.createEdit.fragments

import android.app.Activity
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.text.TextUtils
import android.view.*
import com.elementary.tasks.R
import com.elementary.tasks.core.data.models.Place
import com.elementary.tasks.core.data.models.Reminder
import com.elementary.tasks.core.fragments.AdvancedMapFragment
import com.elementary.tasks.core.interfaces.MapCallback
import com.elementary.tasks.core.interfaces.MapListener
import com.elementary.tasks.core.location.LocationTracker
import com.elementary.tasks.core.utils.*
import com.elementary.tasks.core.views.ActionView
import com.google.android.gms.maps.model.LatLng
import kotlinx.android.synthetic.main.fragment_reminder_location_out.*

/**
 * Copyright 2016 Nazar Suhovich
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
class LocationOutFragment : RadiusTypeFragment() {

    private var mAdvancedMapFragment: AdvancedMapFragment? = null

    private var mTracker: LocationTracker? = null
    private var lastPos: LatLng? = null

    private val mActionListener = object : ActionView.OnActionListener {
        override fun onActionChange(hasAction: Boolean) {
            if (!hasAction) {
                reminderInterface?.setEventHint(getString(R.string.remind_me))
                reminderInterface?.setHasAutoExtra(false, "")
            }
        }

        override fun onTypeChange(isMessageType: Boolean) {
            if (isMessageType) {
                reminderInterface?.setEventHint(getString(R.string.message))
                reminderInterface?.setHasAutoExtra(true, getString(R.string.enable_sending_sms_automatically))
            } else {
                reminderInterface?.setEventHint(getString(R.string.remind_me))
                reminderInterface?.setHasAutoExtra(true, getString(R.string.enable_making_phone_calls_automatically))
            }
        }
    }

    private val mListener = object : MapListener {
        override fun placeChanged(place: LatLng, address: String) {
            lastPos = place
            val location = SuperUtil.getAddress(place.latitude, place.longitude)
            currentLocation.text = location
        }

        override fun onZoomClick(isFull: Boolean) {
            reminderInterface?.setFullScreenMode(isFull)
        }

        override fun onBackClick() {
            val advancedMapFragment = mAdvancedMapFragment ?: return
            if (advancedMapFragment.isFullscreen) {
                advancedMapFragment.isFullscreen = false
                reminderInterface?.setFullScreenMode(false)
            }
            ViewUtils.fadeOutAnimation(mapContainer)
            ViewUtils.fadeInAnimation(specsContainer)
        }
    }

    private fun showPlaceOnMap() {
        val iFace = reminderInterface ?: return
        val item = iFace.reminder
        if (item != null) {
            if (!Reminder.isGpsType(item.type)) return
            val text = item.summary
            val jPlace = item.places[0]
            val latitude = jPlace.latitude
            val longitude = jPlace.longitude
            radius = jPlace.radius
            val advancedMapFragment = mAdvancedMapFragment ?: return
            advancedMapFragment.setMarkerRadius(radius)
            lastPos = LatLng(latitude, longitude)
            advancedMapFragment.addMarker(lastPos, text, true, true, radius)
            mapLocation.text = SuperUtil.getAddress(lastPos!!.latitude, lastPos!!.longitude)
            mapCheck.isChecked = true
        }
    }

    override fun recreateMarker() {
        mAdvancedMapFragment?.recreateMarker(radius)
    }

    override fun prepare(): Reminder? {
        if (super.prepare() == null) return null
        val iFace = reminderInterface ?: return null
        val map = mAdvancedMapFragment ?: return null
        var type = Reminder.BY_OUT
        val isAction = actionView.hasAction()
        if (TextUtils.isEmpty(iFace.summary) && !isAction) {
            iFace.showSnackbar(getString(R.string.task_summary_is_empty))
            return null
        }
        val pos = lastPos
        if (pos == null) {
            iFace.showSnackbar(getString(R.string.you_dont_select_place))
            return null
        }
        var number = ""
        if (isAction) {
            number = actionView.number
            if (TextUtils.isEmpty(number)) {
                iFace.showSnackbar(getString(R.string.you_dont_insert_number))
                return null
            }
            type = if (actionView.type == ActionView.TYPE_CALL) {
                Reminder.BY_OUT_CALL
            } else {
                Reminder.BY_OUT_SMS
            }
        }
        var reminder = iFace.reminder
        if (reminder == null) {
            reminder = Reminder()
        }
        val place = Place(radius, map.markerStyle, pos.latitude, pos.longitude, iFace.summary, number, listOf())
        val places = listOf(place)
        reminder.places = places
        reminder.target = number
        reminder.type = type
        reminder.exportToCalendar = false
        reminder.exportToTasks = false
        reminder.setClear(iFace)
        if (attackDelay.isChecked) {
            val startTime = dateView.dateTime
            reminder.startTime = TimeUtil.getGmtFromDateTime(startTime)
            reminder.eventTime = TimeUtil.getGmtFromDateTime(startTime)
            LogUtil.d(TAG, "EVENT_TIME " + TimeUtil.getFullDateTime(startTime, true, true))
        } else {
            reminder.eventTime = ""
            reminder.startTime = ""
        }
        return reminder
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        setHasOptionsMenu(true)
    }

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
        inflater?.inflate(R.menu.fragment_location_menu, menu)
        super.onCreateOptionsMenu(menu, inflater)
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        when (item!!.itemId) {
            R.id.action_custom_radius -> showRadiusPickerDialog()
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_reminder_location_out, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        val advancedMapFragment = AdvancedMapFragment.newInstance(true, true, true, true,
                prefs.markerStyle, themeUtil.isDark)
        advancedMapFragment.setListener(mListener)
        advancedMapFragment.setCallback(object : MapCallback {
            override fun onMapReady() {
                showPlaceOnMap()
            }
        })
        fragmentManager!!.beginTransaction()
                .replace(mapFrame.id, advancedMapFragment)
                .addToBackStack(null)
                .commit()
        actionView.setListener(mActionListener)
        actionView.setActivity(activity!!)
        actionView.setContactClickListener(View.OnClickListener { selectContact() })

        delayLayout.visibility = View.GONE
        mapContainer.visibility = View.GONE
        attackDelay.setOnCheckedChangeListener { _, isChecked ->
            if (isChecked)
                delayLayout.visibility = View.VISIBLE
            else
                delayLayout.visibility = View.GONE
        }
        mapButton.setOnClickListener {
            if (mapCheck.isChecked) {
                toggleMap()
            }
            mapCheck.isChecked = true
        }
        currentCheck.setOnCheckedChangeListener { _, _ ->
            if (currentCheck.isChecked) {
                mapCheck.isChecked = false
                if (Permissions.checkPermission(activity!!, Permissions.ACCESS_COARSE_LOCATION, Permissions.ACCESS_FINE_LOCATION)) {
                    mTracker = LocationTracker(context) { lat, lng ->
                        lastPos = LatLng(lat, lng)
                        val mLocation = SuperUtil.getAddress(lat, lng)
                        var text = reminderInterface!!.summary
                        if (TextUtils.isEmpty(text)) text = mLocation
                        currentLocation.text = mLocation
                        mAdvancedMapFragment?.addMarker(lastPos, text, true, true, radius)
                    }
                } else {
                    Permissions.requestPermission(activity!!, LOCATION, Permissions.ACCESS_FINE_LOCATION, Permissions.ACCESS_COARSE_LOCATION)
                }
            }
        }
        mapCheck.setOnCheckedChangeListener { _, _ ->
            if (mapCheck.isChecked) {
                currentCheck.isChecked = false
                toggleMap()
                mTracker?.removeUpdates()
            }
        }
        currentCheck.isChecked = true
        editReminder()
    }

    private fun toggleMap() {
        if (mapContainer != null && mapContainer.visibility == View.VISIBLE) {
            ViewUtils.fadeOutAnimation(mapContainer)
            ViewUtils.fadeInAnimation(specsContainer)
        } else {
            ViewUtils.fadeOutAnimation(specsContainer)
            ViewUtils.fadeInAnimation(mapContainer)
            mAdvancedMapFragment?.showShowcase()
        }
    }

    override fun onBackPressed(): Boolean {
        return mAdvancedMapFragment == null || mAdvancedMapFragment!!.onBackPressed()
    }

    private fun editReminder() {
        val iFace = reminderInterface ?: return
        val reminder = iFace.reminder ?: return
        if (reminder.eventTime != "") {
            dateView.setDateTime(reminder.eventTime)
            attackDelay.isChecked = true
        }
        if (reminder.target != "") {
            actionView.setAction(true)
            actionView.number = reminder.target
            if (Reminder.isKind(reminder.type, Reminder.Kind.CALL)) {
                actionView.type = ActionView.TYPE_CALL
            } else if (Reminder.isKind(reminder.type, Reminder.Kind.SMS)) {
                actionView.type = ActionView.TYPE_MESSAGE
            }
        }
    }

    private fun selectContact() {
        if (Permissions.checkPermission(activity!!, Permissions.READ_CONTACTS, Permissions.READ_CALLS)) {
            SuperUtil.selectContact(activity!!, Constants.REQUEST_CODE_CONTACTS)
        } else {
            Permissions.requestPermission(activity!!, CONTACTS, Permissions.READ_CONTACTS, Permissions.READ_CALLS)
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == Constants.REQUEST_CODE_CONTACTS && resultCode == Activity.RESULT_OK) {
            val number = data!!.getStringExtra(Constants.SELECTED_CONTACT_NUMBER)
            actionView.number = number
        }
    }

    override fun onDetach() {
        super.onDetach()
        mTracker?.removeUpdates()
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        actionView.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (grantResults.isEmpty()) return
        when (requestCode) {
            CONTACTS -> if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                selectContact()
            }
            LOCATION -> if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                currentCheck.isChecked = true
            }
            CONTACTS_ACTION -> if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                actionView.setAction(true)
            }
        }
    }

    companion object {

        private const val TAG = "DateFragment"
        private const val CONTACTS = 132
        private const val LOCATION = 202
        const val CONTACTS_ACTION = 133
    }
}
